src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java

```diff
@@ -970,6 +979,407 @@ public void recursivelyDelete(final Project project, final boolean commitIndex)
+                                JOIN PROJECT ON COMPONENT.PROJECT_ID = PROJECT.ID
+                                WHERE PROJECT.UUID = ANY(?)
+                                UNION ALL
+                                SELECT COMPONENT.ID, COMPONENT.PARENT_COMPONENT_ID
+                                FROM COMPONENT
+                                INNER JOIN c_family ON COMPONENT.ID = c_family.PARENT_COMPONENT_ID
+                            )
+                            DELETE FROM COMPONENT
+                            WHERE ID IN (SELECT ID FROM c_family);
+                            """.replace("= ANY(?)", inExpression)
+                        );
+                    }
+                    executeAndCloseWithArray(sqlQuery, queryParameter);
+                }
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """                        
+                    DELETE FROM "COMPONENT" WHERE "PROJECT_ID" IN (
+                        SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                    );
+                    """.replace("= ANY(?)", inExpression)
+                );
+                executeAndCloseWithArray(sqlQuery, queryParameter);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                    DELETE FROM "BOM" WHERE "PROJECT_ID" IN (
+                        SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                    );
+                    """.replace("= ANY(?)", inExpression)
+                );
+                executeAndCloseWithArray(sqlQuery, queryParameter);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                    DELETE FROM "PROJECT_METADATA" WHERE "PROJECT_ID" IN (
+                        SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                    );
+                    """.replace("= ANY(?)", inExpression)
+                );
+                executeAndCloseWithArray(sqlQuery, queryParameter);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                    DELETE FROM "PROJECT_PROPERTY" WHERE "PROJECT_ID" IN (
+                        SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                    );
+                    """.replace("= ANY(?)", inExpression)
+                );
+                executeAndCloseWithArray(sqlQuery, queryParameter);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                    DELETE FROM "PROJECTS_TAGS" WHERE "PROJECT_ID" IN (
+                        SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                    );
+                    """.replace("= ANY(?)", inExpression)
+                );
+                executeAndCloseWithArray(sqlQuery, queryParameter);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                    DELETE FROM "TAG"
+                    WHERE "ID" NOT IN (SELECT "TAG_ID" FROM "PROJECTS_TAGS");
```

Comment:
Tags themselves might be used for other purposes and should not be deleted.

Reply:
Right, that was a bit over-thinking from my side. Now removed.